src/gc.c

```diff
@@ -4070,18 +4070,10 @@ JL_DLLEXPORT jl_value_t *jl_gc_internal_obj_base_ptr(void *p)
         // before the freelist pointer was either live during the last
         // sweep or has been allocated since.
         if (gc_page_data(cell) == gc_page_data(pool->freelist)
-            && (char *)cell < (char *)pool->freelist) {
+            && (char *)cell < (char *)pool->freelist)
             goto valid_object;
-        }
-        else {
-            jl_taggedvalue_t *v = pool->freelist;
-            while (v != NULL) {
-                if (v == cell) {
-                    return NULL;
-                }
-                v = v->next;
-            }
-        }
+        else
+            return NULL;
```

Comment:
OK, after reading the code comments and the code before this, it seems this change is correct. But perhaps a comment could be added to help future reads to understand it? E.g.:
```suggestion
            // already skipped marked or old objects above, so here
            // the age bits are 0, thus the object is on the freelist
            return NULL;
```